httomo/wrappers_class.py

```diff
@@ -0,0 +1,308 @@
+from typing import Dict
+import numpy as np
+import inspect
+from mpi4py.MPI import Comm
+from inspect import signature
+from httomo.utils import print_once
+
+
+gpu_enabled = False
+try:
+    import cupy as xp
+
+    try:
+        xp.cuda.Device(0).compute_capability
+        gpu_enabled = True  # CuPy is installed and GPU is available
+    except xp.cuda.runtime.CUDARuntimeError:
+        import numpy as xp
+
+        print("CuPy is installed but GPU device inaccessible")
+except ImportError:
+    import numpy as xp
+
+    print("CuPy is not installed")
+
+
+class BaseWrapper:
+    """A parent class for all wrappers in httomo that use external modules."""
+
+    def __init__(
+        self, module_name: str, function_name: str, method_name: str, comm: Comm
+    ):
+        self.comm = comm
+        if gpu_enabled:
+            self.num_GPUs = xp.cuda.runtime.getDeviceCount()
+            self.gpu_id = int(comm.rank / comm.size * self.num_GPUs)
```

Comment:
I don't think this is right - assume there are 8 processes spread over 2 nodes with 4 GPUs each. You'd get:

| rank | node | gpu_id |
|------|------|--------|
| 0 | 0 | 0 |
| 1 | 0 | 0 |
| 2 | 0 | 1 |
| 3 | 0 | 1 |
| 4 | 1 | 2 |
| 5 | 1 | 2 |
| 6 | 1 | 3 |
| 7 | 1 | 3 |

Reply:
Sorry I'm not sure how this is calculated. Say you've got 4 GPUs and 4 Processes then size.comm = 4 and you'll get 1,2,3,4 gpu indices for each process. The same distribution will be for a different node. The indexing of GPUs is local to the node, isn't it?

Reply:
I _think_ this is how the table has been constructed...

The values in the `gpu_id` column represent the different values of `self.gpu_id` in the code, which are calculated based on the different values of `comm.rank` for each MPI process via:
```python
self.gpu_id = int(comm.rank / comm.size * self.num_GPUs)
```

The values in the `rank` column are being substituted as `comm.rank` in the code, and then we set `comm.size=8` and `self.num_GPUs=4` for this particular example of
- 2 nodes
- 4 GPUs per node

which comes to 8 processes altogether.

If you plug those values into the line
```python
self.gpu_id = int(comm.rank / comm.size * self.num_GPUs)
```
for each of the 8 MPI processes (ie, changing `comm.rank`), you get the following:
```python
>>> int(0 / 8 * 4)
0
>>> int(1 / 8 * 4)
0
>>> int(2 / 8 * 4)
1
>>> int(3 / 8 * 4)
1
>>> int(4 / 8 * 4)
2
>>> int(5 / 8 * 4)
2
>>> int(6 / 8 * 4)
3
>>> int(7 / 8 * 4)
3
```

Reply:
yes, indexing of GPUs is local, but the comm.size and comm.rank are global. So we need to find which of the ranks are on the same node (just different cores), and establish a local rank between them. So in the table above, it should restart at 0 for the global rank 4.

Reply:
One example to do that is to establish a dict of rank -> hostname + gather it from all processes, and then count the ranks within the same hostname. Like here: https://github.com/ptycho/ptypy/blob/master/ptypy/utils/parallel.py#L746  ... that's rather complex, and I think this would work better:

```python
        local_comm = comm.Split_type(MPI.COMM_TYPE_SHARED)
        if gpu_enabled:
            self.num_GPUs = xp.cuda.runtime.getDeviceCount()
            self.gpu_id = local_comm.rank % self.num_GPUs
```

This uses a local communicator, splitting the global one into those that can communicate with shared memory. That is the same as the ranks on the same node. And within these groups, the rank count starts at 0 again. 

I've added the modulo, so that if for some reason you're putting 8 processes on a node with 4 GPUs, it would wrap around and put the rank 4 back on gpu 0. Another option is to throw an exception for that case.

Reply:
...That's another bit of MPI magic that I've never seen before, thanks for pointing out the existence of `comm.Split_type()`!

Reply:
Just had a short session with @dkazanc - I think it's best to create an MPI utility module in the codebase, which exports `rank`, `size`, `local_rank`, and `local_size`, and have a function in there which determines the local values. So it should be this for the different MPI processes, assuming 2 procs/node:


| size  | 4 | 4 | 4 | 4 |
|----|---|---|---|---|
|  rank  | 0 | 1 | 2 | 3 |
| local_size | 2 | 2 | 2 | 2 |
| local_rank | 0 | 1 | 0 | 1 |

It may be implemented using the `Split_type` call above.

And then everywhere else in the code, you use this utility to get the local ranks and sizes.